docs/pages/UPGRADE-3.0.md

```diff
@@ -0,0 +1,627 @@
+final class Profile extends AggregateRoot
+{
+    // ...
+}
+```
+
+After (Using new basic class):
+```php
+use Patchlevel\EventSourcing\Aggregate\BasicAggregateRoot;
+use Patchlevel\EventSourcing\Attribute\Aggregate;
+
+#[Aggregate('profile')]
+final class Profile extends BasicAggregateRoot
+{
+    // ...
+}
+```
+
+After (Using interface and trait):
+
+```php
+use Patchlevel\EventSourcing\Aggregate\AggregateRootAttributeBehaviour;
+use Patchlevel\EventSourcing\Aggregate\BasicAggregateRoot;
+use Patchlevel\EventSourcing\Attribute\Aggregate;
+
+#[Aggregate('profile')]
+final class Profile implements AggregateRoot
+{
+    use AggregateRootAttributeBehaviour;
+    // ...
+}
+```
+
+#### Aggregate Root ID
+
+We introduced the `#[Id]` Attribute which is used to mark the aggregate root id. This is again a step for more 
+decoupling and enables you to freely design your Aggregates API. The implementation of `AggregateRoot::aggregateRootId` 
+is not needed anymore. 
+
+The aggregate root id now needs to implement the `AggregateRootId` interface instead of just being a string. This 
+enabled us to have more types safety internally. Regarding that we provide a `Uuid` implementation which is using 
+`ramsey/uuid` under the hood and a `CustomId` implementation which is expecting a `string` which could be used if you 
+are not using uuids. You can also provide your own implementation and use that.
+
+Before:
+```php
+use Patchlevel\EventSourcing\Aggregate\AggregateRoot;
+use Patchlevel\EventSourcing\Attribute\Aggregate;
+
+#[Aggregate('profile')]
+final class Profile extends AggregateRoot
+{
+    private ProfileId $id;
+
+    public function aggregateRootId(): string
+    {
+        return $this->id->toString();
+    }
+}
+
+// ProfileId.php
+final class ProfileId
+{
+    private function __construct(private string $id)
+    {
+    }
+
+    public static function fromString(string $id): self
+    {
+        return new self($id);
+    }
+
+    public function toString(): string
+    {
+        return $this->id;
+    }
+}
+```
+
+After:
+```php
+use Patchlevel\EventSourcing\Aggregate\BasicAggregateRoot;
+use Patchlevel\EventSourcing\Attribute\Aggregate;
+use Patchlevel\EventSourcing\Attribute\Id;
+
+#[Aggregate('profile')]
+final class Profile extends BasicAggregateRoot
+{
+    #[Id]
+    private ProfileId $id;
+}
+
+// ProfileId.php
+use Patchlevel\EventSourcing\Aggregate\AggregateRootId;
+
+final class ProfileId implements AggregateRootId
+{
+    private function __construct(private string $id) 
+    {
+    }
+
+    public static function fromString(string $id): self
+    {
+        return new self($id);
+    }
+
+    public function toString(): string
+    {
+        return $this->id;
+    }
+}
+```
+
+### Events
+
+#### Normalizer
+
+The Attribute `#[Normalize]` was removed, use the Normalizer itself instead e.g. `#[IdNormalizer]`. Most of the 
+Normalizers are now located in [patchlevel/hydrator](https://github.com/patchlevel/hydrator).
```

Comment:
It should also be mentioned here that your own normalizer needs the attribute `#[Attribute(Attribute::TARGET_PROPERTY)]`.